README
Some tools to be used with smart cards and PC/SC
================================================

This archive contains some tools usefull for a PC/SC user.
PC/SC lite [1] from MUSCLE [2].

The tools provided are:

- pcsc_scan (Ludovic Rousseau <ludovic.rousseau@free.fr>)
    regularly scans every PC/SC reader connected to the host
    if a card is inserted or removed a "line" is printed

    For example:

     PC/SC device scanner
     V 1.4.5 (c) 2001-2006, Ludovic Rousseau <ludovic.rousseau@free.fr>
     PC/SC lite version: 1.3.1
     0: GemPC410 0 0
     1: GemPC430 0 0
     
     Wed Aug 21 10:08:02 2002
      Reader 0 (GemPC410 0 0)
             Card state: Card removed, 
     
     Wed Aug 21 10:08:09 2002
      Reader 0 (GemPC410 0 0)
             Card state: Card inserted, 
             ATR: 3B 6D 00 FF 00 31 80 71 8E 64 48 D5 02 00 82 90 00
     
     ATR: 3B 6D 00 FF 00 31 80 71 8E 64 48 D5 02 00 82 90 00
     + TS = 3B --> Direct Convention
     + T0 = 6D, Y(1): 0110, K: 13 (historical bytes)
       TB(1) = 00 --> Programming Param P: 0, I: 0
       TC(1) = FF --> Extra guard time: 255
     + Historical bytes: 00 31 80 71 8E 64 48 D5 02 00 82 90 00
       Category indicator byte: 00 (compact TLV data object)
         Tag: 3, len: 1 (card service data byte)
           Card service data byte: 80
             - Application selection: by full DF name
             - EF.DIR and EF.ATR access services: by GET RECORD(s) command
             - Card with MF
         Tag: 7, len: 1 (card capabilities)
           Selection methods: 8E
             - DF selection by full DF name
             - Implicit DF selection
             - Short EF identifier supported
             - Record number supported
         Tag: 6, len: 4 (pre-issuing data)
           Data: 48 D5 02 00
         Mandatory status indicator (3 last bytes)
           LCS (life card cycle): 82 (Proprietary)
           SW: 9000 (Normal processing.)
     
     Possibly identified card:
     3B 6D 00 FF 00 31 80 71 8E 64 48 D5 02 00 82 90 00
             Blue for Business, American Express@Business

     Wed Aug 21 10:09:57 2002
      Reader 0 (GemPC410 0 0)
             Card state: Card removed, 


- ATR_analysis (Christophe Levantis and Ludovic Rousseau)
    Perl script used to parse the smart card ATR.
    This script is called (by default) by pcsc_scan

    This code was originally written by Christophe in August 2000 for a
    Perl wrapper using "TLP driver" (and not PC/SC) to access the
    smartcard.


- smartcard_list.txt (Ludovic Rousseau)
    This list contains ATR of some cards. The list is used by
    ATR_analysis to find a card model corresponding to the ATR.
    I took the initial list from SCEZ by Matthias Bruestle.

    Please send me (<ludovic.rousseau@free.fr>) any ATR and card
    description not included in the list.


- scriptor (Lionel Victor <lionel.victor@unforgettable.com>)
    A Perl script to send commands to a smart card. You can use a file
    containing the commands or stdin.


- gscriptor (Lionel Victor <lionel.victor@unforgettable.com>)
    A Perl script to send commands to a smart card with GTK-based
    graphical interface.


You will need the package pcsc-perl [3] to be able to use scriptor and
gscriptor.

[1] http://pcsclite.alioth.debian.org/
[2] http://linuxnet.com/
[3] http://ludovic.rousseau.free.fr/softwares/pcsc-perl/


Authors:
========

- Lionel VICTOR for scriptor and gscriptor
- Ludovic ROUSSEAU for pcsc_scan, ATR_analysis and debug of scriptor and
  gscriptor
- Christophe LEVANTIS for the original code of ATR_analysis


Licences:
=========

    pcsc-tools
    Copyright (C) 2001       Lionel VICTOR
    Copyright (C) 2001-2008  Ludovic ROUSSEAU

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


History:
========

1.4.17 - 18 August 2010, Ludovic ROUSSEAU
    - 153 new ATRs
    - Allow to build with pcsc-lite >= 1.6.2

1.4.16 - 12 January 2010, Ludovic ROUSSEAU
    - 153 new ATR
    - pcsc_scan.c: check for PnP support at run time instead of using a
      #define
    - ATR_analysis: use curl instead of wget on Darwin
    - gscriptor: ReaderConfig(): escape metacharacters []() in
      the reader name when using reader name as a pattern matching

1.4.15 - 9 January 2009, Ludovic ROUSSEAU
    - 68 new ATR
    - ATR_analysis:
     . truncate the ATR if extra bytes are present (like on Leopard with
     an ATR padded with 0 up to 33 bytes)
     . add value for Di=7 (7816-3:2006 page 19)
     . check if Fi is RFU when calculating baud rate
     . display the max frequency associated with Fi
    - pcsc_scan.c: use "\\?PnP?\Notification" mechanism when possible

1.4.14 - 11 May 2008, Ludovic ROUSSEAU
    - 24 new ATR
    - gscriptor, scriptor: use SCARD_SHARE_SHARED instead of
      SCARD_SHARE_EXCLUSIVE to not lock the card
    - ATR_analysis: add support for ATR with : as byte separator (ATR
      reported by "opensc-tool --atr" for example)

1.4.13 - 23 March 2008, Ludovic ROUSSEAU
    - 29 new ATR
    - pcsc_scan: avoid a bug when the last reader is removed

1.4.12 - 2 February 2008, Ludovic ROUSSEAU
    - 18 new ATR
    - handle SCARD_E_NO_READERS_AVAILABLE when no reader is found

1.4.11 - 30 October 2007, Ludovic ROUSSEAU
    - 69 new ATR
    - use /usr/bin/env perl to get the correct Perl
    - print card description in blue

1.4.10 - 10 August 2007, Ludovic ROUSSEAU
    - 25 new ATR
    - scriptor: correctly display the ISO7816 error code even when the
      line is splitted
    - gscriptor: the line continuation feature was broken

1.4.9 - 1 June 2007, Ludovic ROUSSEAU
    - 69 new ATR
    - pcsc_scan.c: use pcsc_stringify_error() to display a human
      readable error message
    - ATR_analysis: add colors
    - scriptor:  echo the lines read from a file (commands, comments and
      empty lines)

1.4.8 - 26 Novembre 2006, Ludovic ROUSSEAU
    - 14 new ATR
    - pcsc_scan: cleanly fail (instead of looping forever) if the pcscd
      daemon just dies
    - ATR_analysis:
      . add test for TC1=255
      . add case TB1=0: VPP is not electrically connected
      . add the communication speed according to TA1 and a clock at
        3.5712 MHz
      . use %g instead of %.3f to display the cycles/ETU value to have
        31 instead of 31.000

1.4.7 - 4 October 2006, Ludovic ROUSSEAU
    - 10 new ATR
    - pcsc_scan.c: add a couple of fflush(stdout); to force the display
      so that redirection works. Thanks to Johannes Becker for the bug
      report

1.4.6 - 12 August 2006, Ludovic ROUSSEAU
    - 28 new ATRs in smartcard_list.txt
    - scriptor/gscriptor: allow multi-lines input and wrap long output.
      lines ending with \ are to continue on the next line. More easy to
      use a 64K-bytes long APDU.

1.4.5 - 8 May 2006, Ludovic ROUSSEAU
    - 9 new ATR
    - improve ATR_analysis: add code to parse the historical bytes

1.4.4 - 25 Mar 2006, Ludovic ROUSSEAU
    - 4 new ATR
    - improve gscriptor GUI: add icons in the dialog boxes, use button boxes
    - improve ATR_analysis: display and check TCK

1.4.3 - 9 Mar 2006, Ludovic ROUSSEAU
    - the glibc do not accept double free() anymore and stops the
      process instead "*** glibc detected *** double free or corruption
      (fasttop): 0x0804b070 ***". So be sure not to call free() a second
      time on the same pointer.

1.4.2 - 7 Mar 2006, Ludovic ROUSSEAU
    - 58 new ATR in smartcard_list.txt
    - pcsc_scan: add color output support
    - add a freedesktop.org compatible menu entry for gscriptor (thanks
      to Ville Skyttä)

1.4.1 - 29 May 2005, Ludovic ROUSSEAU
    - 42 new ATR in smartcard_list.txt
    - ATR_analysis:
     . accept 3B A7 00 40 18 80 65 A2 08 01 01 52 and not just '3B A7 00
       40 18 80 65 A2 08 01 01 52' as command line argument
     . check ~/.smartcard_list.txt for known ATR
     . add instructions to download and install a new version of the list
    - gscriptor: if only one reader is found use it unconditionally

1.4.0 - 6 August 2004, Ludovic ROUSSEAU
    - 13 new ATR in smartcard_list.txt
    - rewrite of gscriptor using libgtk2-perl
    - if the command does not contains spaces (00A4030000) we expand it

1.3.4 - 4 July 2004, Ludovic ROUSSEAU
    - 5 new ATR in smartcard_list.txt
    - pcsc_scan.c: LPSTR -> LPTSTR for pcsc-lite 1.2.9 beta4
    - scriptor, gscriptor: only send the 5 last characters of the
      response to Chipcard::PCSC::Card::ISO7816Error()
      The SW was sometimes not correctly parsed.
    - Makefile: use 'pkg-config libpcsclite [--cflags|--libs]'

1.3.3 - 2 April 2004, Ludovic ROUSSEAU
    - use a dynamic table for readers to avoid any use of the deprecated
      PCSCLITE_MAX_CHANNELS constant
    - 11 new ATR in smartcard_list.txt
    - add support of PCSC MacOS X framework

1.3.2 - 17 December 2003, Ludovic ROUSSEAU
    - scriptor: by default let pcscd select the protocol among T=0 and T=1 
      . print the protocol actually used
      . print textual status word according to ISO 7816
    - gscriptor: change menu accelerator for Run so it can be used even
      if a text Text widget is selected
      . use paned windows to resize Script and Result frames
      . print textual status word according to ISO 7816
    - 16 new ATR in smartcard_list.txt

1.3.1 - 29 Octobre 2003, Ludovic ROUSSEAU
    - ATR_analysis: only ask to send me the ATR and card description if
      it is a microprocessor card. I am not (yet) interested in a list
      of memory cards.
    - scriptor, scriptor.1p: add -p argument to specify the ISO 7816
      protocol to use (T=0 or T=1)
    - 26 new ATR in smartcard_list.txt

1.3.0 - 31 May 2003, Ludovic ROUSSEAU
    - pcsc_scan.c: detect reader insertion/removal (USB) and adjust the
      reader list accordingly
    - 4 new ATR in smartcard_list.txt

1.2.5 - 25 May 2003, Ludovic ROUSSEAU
    - use new naming scheme from pcsc-perl Chipcard::PCSC instead of
      PCSC
    - 32 new ATR in smartcard_list.txt

1.2.4 - 28 Jan 2003, Ludovic ROUSSEAU
    - 20 new ATRs in the database
    - Makefile: use PCSCBASE to find libpcsclite library

1.2.3 - 16 Nov 2002, Ludovic ROUSSEAU
    - Add the URL of the latest smartcard_list.txt file in the error
      message printed if the ATR is not already known
    - use an updated version of smartcard_list.txt

1.2.2 - 15 Oct 2002, Ludovic ROUSSEAU
    - correct some compilation problems. See Makefile if you have a xBSD
      system
    - use an updated version of smartcard_list.txt. Thanks to all the
      contributors.

1.2.1 - 27 Aug 2002, Ludovic ROUSSEAU
    - add ATR to smartcard_list.txt
    - make the ATR match case insensitive

1.2.0 - 21 Aug 2002,  Ludovic ROUSSEAU
    - creation of smartcard_list.txt
    - exploitation of the ATR list by ATR_analysis

1.1.0 - 14 Jun 2002, Ludovic ROUSSEAU
    - include ATR_analysis and support for it in pcsc_scan
    - modify pcsc_scan.c to listen to all the readers at once (Lionel)

1.0.4 - 6 Mar 2002, Ludovic ROUSSEAU
    - gscriptor: case insensitive on "reset" command
    - scriptor: case insensitive on "exit" and "reset" commands
    - add the history, authors and licence in this README

1.0.3 - 8 Nov 2001, Ludovic ROUSSEAU
    - pcsc_scan.c: set the wait time to 0 to get all the events
    - Makefile: add automatic rules to compress man pages and other
      modifications

1.0.2 - 22 Oct 2001, Ludovic ROUSSEAU
    - add manpages
    - pcsc_scan.c: go to the next reader _after_ printing information
      for the current one

1.0.1 - 18 Oct 2001, Ludovic ROUSSEAU

$Id: README 5162 2010-08-18 21:42:46Z rousseau $

 vim:ts=8:


Changelog
2010-08-18  rousseau

	* [r5161] pcsc_scan.c: Use SCARD_READERSTATE instead of
	  SCARD_READERSTATE_A since it is no more defined in pcsc-lite >=
	  1.6.2

2010-08-17  rousseau

	* [r5158] smartcard_list.txt: 1 new ATR

2010-08-14  rousseau

	* [r5131] smartcard_list.txt: add a card description

2010-08-11  rousseau

	* [r5122] smartcard_list.txt: 2 new ATRs

2010-07-24  rousseau

	* [r5068] smartcard_list.txt: 5 new ATRs

2010-07-08  rousseau

	* [r5064] smartcard_list.txt: 1 new ATR

2010-07-02  rousseau

	* [r5063] smartcard_list.txt: 1 new ATR

2010-07-01  rousseau

	* [r5053] smartcard_list.txt: update card description

2010-06-30  rousseau

	* [r5049] smartcard_list.txt: update card description

2010-06-29  rousseau

	* [r5048] smartcard_list.txt: 1 new ATR

2010-06-27  rousseau

	* [r5042] smartcard_list.txt: remove space at end of line
	* [r5041] smartcard_list.txt: 2 new ATRs

2010-06-22  rousseau

	* [r5035] smartcard_list.txt: 1 new ATR

2010-06-20  rousseau

	* [r5018] smartcard_list.txt: 1 new ATR

2010-06-12  rousseau

	* [r4990] smartcard_list.txt: Add an ATR description

2010-06-10  rousseau

	* [r4989] smartcard_list.txt: 1 new ATR

2010-06-08  rousseau

	* [r4984] smartcard_list.txt: 3 new ATRs

2010-06-04  rousseau

	* [r4976] smartcard_list.txt: 1 new ATR

2010-05-28  rousseau

	* [r4966] smartcard_list.txt: 2 new ATRs

2010-05-25  rousseau

	* [r4962] smartcard_list.txt: 1 new ATR

2010-05-24  rousseau

	* [r4959] smartcard_list.txt: 1 new ATR

2010-05-20  rousseau

	* [r4958] smartcard_list.txt: 1 new ATR

2010-05-18  rousseau

	* [r4955] smartcard_list.txt: add a card description

2010-05-15  rousseau

	* [r4942] smartcard_list.txt: 2 new ATRs

2010-05-08  rousseau

	* [r4935] smartcard_list.txt: 1 new ATR

2010-05-07  rousseau

	* [r4929] smartcard_list.txt: 1 new ATR

2010-05-05  rousseau

	* [r4922] smartcard_list.txt: 1 new ATR

2010-05-04  rousseau

	* [r4919] smartcard_list.txt: 2 new ATRs

2010-04-28  rousseau

	* [r4898] smartcard_list.txt: update 1 ATR

2010-04-25  rousseau

	* [r4895] smartcard_list.txt: 1 new ATR

2010-04-24  rousseau

	* [r4892] smartcard_list.txt: 3 new ATRs

2010-04-22  rousseau

	* [r4891] smartcard_list.txt: 1 new ATR
	* [r4890] smartcard_list.txt: 1 new ATR

2010-04-20  rousseau

	* [r4887] smartcard_list.txt: 1 new ATR

2010-04-17  rousseau

	* [r4880] smartcard_list.txt: 2 new ATRs

2010-04-09  rousseau

	* [r4868] smartcard_list.txt: 4 new ATRs

2010-04-06  rousseau

	* [r4853] smartcard_list.txt: 1 new ATR

2010-04-02  rousseau

	* [r4852] smartcard_list.txt: 68 new ATRs families from the PCSC
	  standard, Part 3, Chapter 3.1.3.2.3.2, and Part 3 Supplemental
	  Document

2010-03-25  rousseau

	* [r4842] smartcard_list.txt: 1 new ATR

2010-03-24  rousseau

	* [r4841] smartcard_list.txt: 1 new ATR

2010-03-23  rousseau

	* [r4840] smartcard_list.txt: 1 new ATR

2010-03-21  rousseau

	* [r4837] smartcard_list.txt: 1 new ATR

2010-03-19  rousseau

	* [r4835] smartcard_list.txt: 1 new ATR

2010-03-17  rousseau

	* [r4828] smartcard_list.txt: use . instead of ? in a pattern
	  matching
	* [r4827] smartcard_list.txt: generalize an ATR
	* [r4826] smartcard_list.txt: 3 new ATRs

2010-03-09  rousseau

	* [r4811] smartcard_list.txt: 1 new ATR

2010-03-06  rousseau

	* [r4798] smartcard_list.txt: 1 new ATR
	* [r4797] smartcard_list.txt: 2 new ATRs

2010-03-04  rousseau

	* [r4794] smartcard_list.txt: 3 new ATRs

2010-03-03  rousseau

	* [r4787] smartcard_list.txt: 1 new ATR

2010-03-02  rousseau

	* [r4786] smartcard_list.txt: 1 new ATR
	* [r4785] smartcard_list.txt: correct an ATR

2010-02-26  rousseau

	* [r4782] smartcard_list.txt: 1 new ATR

2010-02-23  rousseau

	* [r4774] smartcard_list.txt: 1 new ATR
	* [r4770] smartcard_list.txt: 4 new ATRs

2010-02-22  rousseau

	* [r4767] smartcard_list.txt: avoid french in card descriptions
	  
	  Thanks to Martin Paljak for the bug report

2010-02-20  rousseau

	* [r4764] smartcard_list.txt: 3 new ATRs

2010-02-12  rousseau

	* [r4763] smartcard_list.txt: 2 new ATRs

2010-02-11  rousseau

	* [r4752] smartcard_list.txt: 2 new ATRs

2010-02-03  rousseau

	* [r4706] smartcard_list.txt: 1 new ATR

2010-01-28  rousseau

	* [r4693] smartcard_list.txt: 1 new ATR

2010-01-26  rousseau

	* [r4688] smartcard_list.txt: relax an ATR matching pattern

2010-01-24  rousseau

	* [r4687] smartcard_list.txt: 2 new ATRs

2010-01-22  rousseau

	* [r4686] smartcard_list.txt: 1 new ATR

2010-01-20  rousseau

	* [r4679] smartcard_list.txt: 1 new ATR

2010-01-19  rousseau

	* [r4678] smartcard_list.txt: 1 new ATR
	* [r4677] smartcard_list.txt: 1 new ATR

2010-01-13  rousseau

	* [r4664] smartcard_list.txt: 1 new ATR
	* [r4663] smartcard_list.txt: generalize the Feitian PKI ATR

2010-01-12  rousseau

	* [r4661] README: release 1.4.16
	* [r4660] create_distrib.sh: ignore ".svn" instead of "CVS" files
	* [r4659] create_distrib.sh: use Changelog instead of (default)
	  ChangeLog
	* [r4658] create_distrib.sh: use svn2cl instead of rcs2log now we
	  are on subversion
	* [r4657] gscriptor: update copyright date
	* [r4656] gscriptor: ReaderConfig(): escape metacharacters []() in
	  the reader name when using reader name as a pattern matching

2010-01-09  rousseau

	* [r4648] smartcard_list.txt: 1 new ATR
	* [r4647] smartcard_list.txt: 1 new ATR

2010-01-08  rousseau

	* [r4643] smartcard_list.txt: 9 new ATRs

2010-01-06  rousseau

	* [r4642] smartcard_list.txt: simplify the Austrian Quick E-purse
	  cards. Every quick card's identification number is BCD encoded in
	  the four contiguous free bytes in the ATR regular expression

2010-01-05  rousseau

	* [r4641] smartcard_list.txt: 5 new ATRs

2010-01-01  rousseau

	* [r4626] smartcard_list.txt: 1 new ATR

2009-12-31  rousseau

	* [r4624] smartcard_list.txt: 11 new ATRs

2009-12-20  rousseau

	* [r4623] smartcard_list.txt: 1 new ATR

2009-12-18  rousseau

	* [r4622] smartcard_list.txt: reformat

2009-12-12  rousseau

	* [r4609] smartcard_list.txt: 1 new ATR


pcsc_scan.c
/*
    Scan and print all the PC/SC devices available
    Copyright (C) 2001-2009  Ludovic Rousseau <ludovic.rousseau@free.fr>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
*/

/* $Id: pcsc_scan.c 5161 2010-08-18 21:36:11Z rousseau $ */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <string.h>

#ifdef __APPLE__
#include <PCSC/wintypes.h>
#include <PCSC/winscard.h>
#else
#include <winscard.h>
#endif

#ifndef TRUE
#define TRUE 1
#define FALSE 0
#endif

#define TIMEOUT 1000	/* 1 second timeout */

/* command used to parse (on screen) the ATR */
#define ATR_PARSER "ATR_analysis"

#ifndef SCARD_E_NO_READERS_AVAILABLE
#define SCARD_E_NO_READERS_AVAILABLE 0x8010002E
#endif

#define test_rv(fct, rv, hContext) \
do { \
	if (rv != SCARD_S_SUCCESS) \
	{ \
		printf("%s%s: %s%s\n", red, fct, pcsc_stringify_error(rv), color_end); \
		(void)SCardReleaseContext(hContext); \
		exit(-1); \
	} \
} while(0)

void usage(void)
{
	printf("usage: pcsc_scan [-n] [-V] [-h] [-t]\n");
	printf("  -n : no ATR analysis\n");
	printf("  -V : print version number\n");
	printf("  -h : this help\n");
	printf("  -t : test\n");
} /* usage */

int main(int argc, char *argv[])
{
	int current_reader;
	LONG rv;
	SCARDCONTEXT hContext;
	SCARD_READERSTATE *rgReaderStates_t = NULL;
	SCARD_READERSTATE rgReaderStates[1];
	DWORD dwReaders = 0, dwReadersOld;
	DWORD timeout;
	LPSTR mszReaders = NULL;
	char *ptr, **readers = NULL;
	int nbReaders, i;
	char atr[MAX_ATR_SIZE*3+1];	/* ATR in ASCII */
	char atr_command[sizeof(atr)+sizeof(ATR_PARSER)+2+1];
	int opt;
	int analyse_atr = TRUE;
	char *blue = "";
	char *red = "";
	char *magenta = "";
	char *color_end = "";
	int pnp = TRUE;

	printf("PC/SC device scanner\n");
	printf("V " VERSION " (c) 2001-2009, Ludovic Rousseau <ludovic.rousseau@free.fr>\n");
        printf(" 	  ver:0.1, Bjoern Heller <tec@hellercom.de>\n");
	printf("Compiled with PC/SC lite version: " PCSCLITE_VERSION_NUMBER "\n");

	while ((opt = getopt(argc, argv, "Vhn")) != EOF)
	{
		switch (opt)
		{
			case 'n':
				analyse_atr = FALSE;
				break;

			case 'V':
				/* the version number is printed by default */
				return 1;
				break;

			case 'h':
				default:
				usage();
				return 1;
				break;

			case 't':
				printf("test\n");
				return 1;
				break;
		}
	}

	if (argc - optind != 0)
	{
		usage();
		return 1;
	}

	/* check if terminal supports color */
	{
		const char *terms[] = { "linux", "xterm", "xterm-color", "Eterm", "rxvt", "rxvt-unicode" };
		char *term;

		term = getenv("TERM");
		if (term)
		{
			int i;

			/* for each known color terminal */
			for (i = 0; i < sizeof(terms) / sizeof(terms[0]); i++)
			{
				/* we found a supported term? */
				if (0 == strcmp(terms[i], term))
				{
					blue = "\33[34m";
					red = "\33[31m";
					magenta = "\33[35m";
					color_end = "\33[0m";
					break;
				}
			}
		}
	}

	rv = SCardEstablishContext(SCARD_SCOPE_SYSTEM, NULL, NULL, &hContext);
	test_rv("SCardEstablishContext", rv, hContext);

	rgReaderStates[0].szReader = "\\\\?PnP?\\Notification";
	rgReaderStates[0].dwCurrentState = SCARD_STATE_UNAWARE;

	rv = SCardGetStatusChange(hContext, 0, rgReaderStates, 1);
	if (rgReaderStates[0].dwEventState && SCARD_STATE_UNKNOWN)
	{
		printf("PnP reader name not supported. Using polling.\n");
		pnp = FALSE;
	}

get_readers:
	/* free memory possibly allocated in a previous loop */
	if (NULL != readers)
	{
		free(readers);
		readers = NULL;
	}

	if (NULL != rgReaderStates_t)
	{
		free(rgReaderStates_t);
		rgReaderStates_t = NULL;
	}

	/* Retrieve the available readers list.
	 *
	 * 1. Call with a null buffer to get the number of bytes to allocate
	 * 2. malloc the necessary storage
	 * 3. call with the real allocated buffer
	 */
	printf("%sScanning present readers...%s\n", red, color_end);
	rv = SCardListReaders(hContext, NULL, NULL, &dwReaders);
	if (rv != SCARD_E_NO_READERS_AVAILABLE)
		test_rv("SCardListReaders", rv, hContext);

	dwReadersOld = dwReaders;

	/* if non NULL we came back so free first */
	if (mszReaders)
	{
		free(mszReaders);
		mszReaders = NULL;
	}

	mszReaders = malloc(sizeof(char)*dwReaders);
	if (mszReaders == NULL)
	{
		printf("malloc: not enough memory\n");
		return 1;
	}

	*mszReaders = '\0';
	rv = SCardListReaders(hContext, NULL, mszReaders, &dwReaders);

	/* Extract readers from the null separated string and get the total
	 * number of readers */
	nbReaders = 0;
	ptr = mszReaders;
	while (*ptr != '\0')
	{
		ptr += strlen(ptr)+1;
		nbReaders++;
	}

	if (SCARD_E_NO_READERS_AVAILABLE == rv || 0 == nbReaders)
	{
		printf("%sWaiting for the first reader...%s", red, color_end);
		fflush(stdout);

		if (pnp)
		{
			rgReaderStates[0].szReader = "\\\\?PnP?\\Notification";
			rgReaderStates[0].dwCurrentState = SCARD_STATE_UNAWARE;

			rv = SCardGetStatusChange(hContext, INFINITE, rgReaderStates, 1);
			test_rv("SCardGetStatusChange", rv, hContext);
		}
		else
		{
			rv = SCARD_S_SUCCESS;
			while ((SCARD_S_SUCCESS == rv) && (dwReaders == dwReadersOld))
			{
				rv = SCardListReaders(hContext, NULL, NULL, &dwReaders);
				if (SCARD_E_NO_READERS_AVAILABLE == rv)
					rv = SCARD_S_SUCCESS;
				sleep(1);
			}
		}
		printf("found one\n");
		goto get_readers;
	}
	else
		test_rv("SCardListReader", rv, hContext);

	/* allocate the readers table */
	readers = calloc(nbReaders+1, sizeof(char *));
	if (NULL == readers)
	{
		printf("Not enough memory for readers table\n");
		return -1;
	}

	/* fill the readers table */
	nbReaders = 0;
	ptr = mszReaders;
	while (*ptr != '\0')
	{
		printf("%s%d: %s%s\n", blue, nbReaders, ptr, color_end);
		readers[nbReaders] = ptr;
		ptr += strlen(ptr)+1;
		nbReaders++;
	}

	/* allocate the ReaderStates table */
	rgReaderStates_t = calloc(nbReaders+1, sizeof(* rgReaderStates_t));
	if (NULL == rgReaderStates_t)
	{
		printf("Not enough memory for readers states\n");
		return -1;
	}

	/* Set the initial states to something we do not know
	 * The loop below will include this state to the dwCurrentState
	 */
	for (i=0; i<nbReaders; i++)
	{
		rgReaderStates_t[i].szReader = readers[i];
		rgReaderStates_t[i].dwCurrentState = SCARD_STATE_UNAWARE;
	}
	rgReaderStates_t[nbReaders].szReader = "\\\\?PnP?\\Notification";
	rgReaderStates_t[nbReaders].dwCurrentState = SCARD_STATE_UNAWARE;

	/* Wait endlessly for all events in the list of readers
	 * We only stop in case of an error
	 */
	if (pnp)
	{
		timeout = INFINITE;
		nbReaders++;
	}
	else
		timeout = TIMEOUT;
	rv = SCardGetStatusChange(hContext, timeout, rgReaderStates_t, nbReaders);
	while ((rv == SCARD_S_SUCCESS) || (rv == SCARD_E_TIMEOUT))
	{
		if (pnp)
		{
			if (rgReaderStates_t[nbReaders-1].dwEventState &
					SCARD_STATE_CHANGED)
				goto get_readers;
		}
		else
		{
			/* A new reader appeared? */
			if ((SCardListReaders(hContext, NULL, NULL, &dwReaders)
						== SCARD_S_SUCCESS) && (dwReaders != dwReadersOld))
				goto get_readers;
		}

		/* Now we have an event, check all the readers in the list to see what
		 * happened */
		for (current_reader=0; current_reader < nbReaders; current_reader++)
		{
			time_t t;

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_CHANGED)
			{
				/* If something has changed the new state is now the current
				 * state */
				rgReaderStates_t[current_reader].dwCurrentState =
					rgReaderStates_t[current_reader].dwEventState;
			}
			else
				/* If nothing changed then skip to the next reader */
				continue;

			/* From here we know that the state for the current reader has
			 * changed because we did not pass through the continue statement
			 * above.
			 */

			/* Timestamp the event as we get notified */
			t = time(NULL);

			/* Specify the current reader's number and name */
			printf("\n%s Reader %d: %s%s%s\n", ctime(&t), current_reader,
				magenta, rgReaderStates_t[current_reader].szReader,
				color_end);

			/* Dump the full current state */
			printf("  Card state: %s", red);

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_IGNORE)
				printf("Ignore this reader, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_UNKNOWN)
			{
				printf("Reader unknown\n");
				goto get_readers;
			}

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_UNAVAILABLE)
				printf("Status unavailable, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_EMPTY)
				printf("Card removed, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_PRESENT)
				printf("Card inserted, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_ATRMATCH)
				printf("ATR matches card, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_EXCLUSIVE)
				printf("Exclusive Mode, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_INUSE)
				printf("Shared Mode, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_MUTE)
				printf("Unresponsive card - more than one card on reader?, ");

			printf("%s\n", color_end);

			/* force display */
			fflush(stdout);

			/* Also dump the ATR if available */
			if (rgReaderStates_t[current_reader].cbAtr > 0)
			{
				printf("  ATR: ");

				if (rgReaderStates_t[current_reader].cbAtr)
				{
					for (i=0; i<rgReaderStates_t[current_reader].cbAtr; i++)
						sprintf(&atr[i*3], "%02X ",
							rgReaderStates_t[current_reader].rgbAtr[i]);

					atr[i*3-1] = '\0';
				}
				else
					atr[0] = '\0';

				printf("%s%s%s\n\n", magenta, atr, color_end);

				/* force display */
				fflush(stdout);

				if (analyse_atr)
				{
					sprintf(atr_command, ATR_PARSER " '%s'", atr);
					if (system(atr_command))
						perror(atr_command);
				}
			}
		} /* for */

		rv = SCardGetStatusChange(hContext, timeout, rgReaderStates_t,
			nbReaders);
	} /* while */

	/* If we get out the loop, GetStatusChange() was unsuccessful */
	test_rv("SCardGetStatusChange", rv, hContext);
	
	/* We try to leave things as clean as possible */
	rv = SCardReleaseContext(hContext);
	test_rv("SCardReleaseContext", rv, hContext);

	/* free memory possibly allocated */
	if (NULL != readers)
		free(readers);
	if (NULL != rgReaderStates_t)
		free(rgReaderStates_t);

	return 0;
} /* main */

